lucene/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTruncQuery.java

```diff
@@ -58,62 +70,16 @@ protected boolean matchingChar(char c) {
     return (c != unlimited) && (c != mask);
   }
 
-  protected void appendRegExpForChar(char c, StringBuilder re) {
-    if (c == unlimited) re.append(".*");
-    else if (c == mask) re.append(".");
-    else re.append(c);
-  }
-
-  protected void truncatedToPrefixAndPattern() {
-    int i = 0;
-    while ((i < truncated.length()) && matchingChar(truncated.charAt(i))) {
-      i++;
-    }
-    prefix = truncated.substring(0, i);
-    prefixRef = new BytesRef(prefix);
-
-    StringBuilder re = new StringBuilder();
-    while (i < truncated.length()) {
-      appendRegExpForChar(truncated.charAt(i), re);
-      i++;
-    }
-    pattern = Pattern.compile(re.toString());
-  }
-
   @Override
   public void visitMatchingTerms(IndexReader reader, String fieldName, MatchingTermVisitor mtv)
       throws IOException {
-    int prefixLength = prefix.length();
     Terms terms = MultiTerms.getTerms(reader, fieldName);
     if (terms != null) {
-      Matcher matcher = pattern.matcher("");
-      try {
-        TermsEnum termsEnum = terms.iterator();
-
-        TermsEnum.SeekStatus status = termsEnum.seekCeil(prefixRef);
-        BytesRef text;
-        if (status == TermsEnum.SeekStatus.FOUND) {
-          text = prefixRef;
-        } else if (status == TermsEnum.SeekStatus.NOT_FOUND) {
-          text = termsEnum.term();
-        } else {
-          text = null;
-        }
+      TermsEnum termsEnum = terms.intersect(compiled, prefixRef);
```

Comment:
@magibney This looks much cleaner! I do have a few noob questions perhaps you can help me with!
1. It appears that `compiled` here contains the full wildcard pattern. The `prefixRef` is a way to "skip" entries in the intersect until it hits entries that could potentially match? When i looked at the javadoc, i'm a bit confused, especially this: `Note that the provided startTerm must be accepted by the automaton.`. So for example, if I have pattern `ab*yz`, that means the start term should be accepted, ie `abyz` is okay, but `ab` is not? In this case, we are passing the prefix as start term but using the full pattern might not match that? 
2. In the previous implementation, it also breaks the pattern into prefix and then anything after the wildcard, but the `matcher` excludes the prefix part, and prefix matching is simply a `StringHelper.startsWith` check? Should we something similar assuming that's certain performance advantage (I don't know if there is TBH 😅 ) to exclude prefix from matcher/automaton?

Reply:
1. is a good catch. I actually _don't_ think that requirement has any practical basis (anymore, at least? or for our context anyway?). But I think the prefix skipping implemented at the PostingsFormat level is actually efficient enough that leaving startTerm=null should be perfectly fine, so I'll just do that so that we can comply with the javadoc instructions.
2. is a great question, but the answer I think follows from the answer to 1.^ : the optimizations are built in at the level of the PostingsFormat, so it's not necessary to split the prefix part out the way the old impl did.

Reply:
481faff326914c684e83f726b3b32389d2bcde2d addresses this. Even more simplification!